source/common/quic/envoy_quic_proof_verifier.cc

```diff
@@ -23,11 +83,62 @@ quic::QuicAsyncStatus EnvoyQuicProofVerifier::VerifyCertChain(
   }
   ENVOY_BUG(!verify_context->isServer(), "Client certificates are not supported in QUIC yet.");
 
-  if (doVerifyCertChain(hostname, port, certs, ocsp_response, cert_sct, context, error_details,
-                        out_alert, std::move(callback))) {
-    *details = std::make_unique<CertVerifyResult>(true);
-    return quic::QUIC_SUCCESS;
+  if (!Runtime::runtimeFeatureEnabled("envoy.reloadable_features.tls_async_cert_validation")) {
+    if (doVerifyCertChain(hostname, port, certs, ocsp_response, cert_sct, context, error_details,
+                          out_alert, std::move(callback))) {
+      *details = std::make_unique<CertVerifyResult>(true);
+      return quic::QUIC_SUCCESS;
+    }
+    *details = std::make_unique<CertVerifyResult>(false);
+    return quic::QUIC_FAILURE;
   }
+
+  bssl::UniquePtr<STACK_OF(X509)> cert_chain(sk_X509_new_null());
+  for (const auto& cert_str : certs) {
+    bssl::UniquePtr<X509> cert = parseDERCertificate(cert_str, error_details);
+    if (!cert || !bssl::PushToStack(cert_chain.get(), std::move(cert))) {
+      return quic::QUIC_FAILURE;
+    }
+  }
+  std::unique_ptr<quic::CertificateView> cert_view =
+      quic::CertificateView::ParseSingleCertificate(certs[0]);
+  ASSERT(cert_view != nullptr);
+  int sign_alg = deduceSignatureAlgorithmFromPublicKey(cert_view->public_key(), error_details);
+  if (sign_alg == 0) {
+    return quic::QUIC_FAILURE;
+  }
+
+  auto* envoy_callback =
+      new QuicValidateResultCallback(verify_context->dispatcher(), std::move(callback), hostname);
+  // We down cast rather than add verifyCertChain to Envoy::Ssl::Context because
```

Comment:
Actually, after reading the rest of the PR, it may be less-bad to add the method to the interface, but pass the STACK_OF(X509) as a `void*` to keep the ssl-internals out of the interface and cast that back to the correct type in the implementation of the method.

Reply:
Sounds reasonable